---
title: "TNBC"
output: html_notebook
---

```{r}
# pacman is a "package manager" that defines the p_load function.
# p_load automatically installs a library if it is not already installed.
# This often, but doesn't always, works.
if(!require("pacman")) {
  install.packages("pacman")
}
suppressPackageStartupMessages(library(pacman)) 
if(!require(Seurat)) remotes::install_github("satijalab/seurat", "seurat5", quiet = TRUE)
# install_version("igraph", version = "1.6.0", repos = "http://cran.us.r-project.org")
suppressPackageStartupMessages(library(Seurat))
# Install presto for fast version of wilcoxon rank sum test in Seurat
if(!require(presto)) devtools::install_github('immunogenomics/presto')
suppressPackageStartupMessages(library(presto))
suppressPackageStartupMessages(p_load(ggplot2))
suppressPackageStartupMessages(p_load(dplyr))
suppressPackageStartupMessages(p_load(plyr))
suppressPackageStartupMessages(p_load(cowplot))
suppressPackageStartupMessages(library(hdf5r))
if(!require("spacexr")) {
  devtools::install_github("dmcable/spacexr", build_vignettes = FALSE)
  suppressPackageStartupMessages(p_load(spacexr))
}
```


Setup a parallel execution environment.

```{r}
suppressPackageStartupMessages(p_load(foreach))
suppressPackageStartupMessages(p_load(parallel))
num.cores <- detectCores()
if(!is.na(num.cores) && (num.cores > 1)) {
  suppressPackageStartupMessages(p_load("doMC"))
  # num.cores <- min(num.cores,6)
  cat(paste("Registering ", num.cores-1, " cores.\n", sep=""))
  registerDoMC(cores=(num.cores-1))
  options(mc.cores=num.cores-1)
}

```

```{r}
# Load some utility functions
url <- "https://raw.githubusercontent.com/TheJacksonLaboratory/cancer-cube-st-pilot-ffpe-ff/main/utils.R"
source(url)
url <- "https://raw.githubusercontent.com/TheJacksonLaboratory/cancer-cube-st-pilot-ffpe-ff/main/plotting-utils.R"
source(url)
```

```{r}
# Initialization
raw.data.dir <- "/projects/activities/jaxcc/CATch-project/raw-data/visium/tnbc/mouse"
# Should define an analysis directory here
```

```{r}
# Define the spaceranger directories for each of the samples
sample.names <- list.files(raw.data.dir)
spaceranger_dirs <- paste0(raw.data.dir, "/", sample.names, "/spaceranger/")
names(spaceranger_dirs) <- sample.names
names(sample.names) <- sample.names
```

```{r}
# Read in the filtered data
filtered.objs <- create.visium.seurat.objects(spaceranger_dirs, filter.spots=TRUE)

# Read in the unfiltered data
unfiltered.objs <- create.visium.seurat.objects(spaceranger_dirs, filter.spots=FALSE)
```

```{r}
# Plot the H&E, UMI counts, and feature counts for each of the _unfiltered_ samples.
# First goal here is to ensure that few reads are outside the tissue.
# Note: plotting unnormalized counts! This may be misleading. We may want log normalized or SCtransform'ed counts.
# But this suffices to show reads outside the tissue.
# Also plotting some genes of interest:
# Actb = beta actin: housekeeping gene, should be broadly expressed
# Vim = Vimentin: expressed on mesenchymal cells (including fibroblasts)
# Krt5 = Keratin 5: a basal keratin
# Krt8 = Keratin 8: a luminal keratin
# Ptprc = Cd45: pan-immune marker
# Cd8a: cytotoxic CD8 T cells
# Having seen that they are, we can do all subsequent analysis on the filtered.objs.
plots <- 
  llply(sample.names,
        .fun = function(sample.name) {
          print(sample.name)
          g <- plot.features(unfiltered.objs[[sample.name]], features=c("Actb", "Vim", "Krt5", "Krt8", "Ptprc", "Cd8a"), feature.names=c("Actb", "Vim", "Krt5", "Krt8", "Cd45", "Cd8a"), include.hne=TRUE, include.umi.cnts=TRUE, include.feature.cnts=TRUE)
          g <- add.title.to.plot(g, title=sample.name, rel_heights=c(0.1,1))
          g
        })
```

```{r}
for(g in plots) {
  print(g)
}
```

